MetaMask/metamask-extension
# Adding `Box` component in TS to component-library

## Explanation
Currently, the `Box` component, a utility component used as a base for all `component-library` components as well as probably the most used component in the code base is still built in JavaScript. A types `box.d.ts` file was created as a workaround for allowing the `Box` component to be used in TypeScript files. However, through many community contributor TS migration PRs we have discovered that the polymorphic `as` prop is making all components difficult to migrate. This currently blocks all typescript migration and proper use of the Box component in TypeScript. 

To resolve this issue this PR creates a new `Box` component built in the `component-library` folder that:

-  ✅  Is built in TypeScript
-  ✅  Allows for correct typing of polymorphic props
-  ✅  Fixes https://github.com/MetaMask/metamask-extension/issues/16976
-  ✅  Allows all design system components to be imported from the same place e.g. `import { Box, Text, Button } from '../../component-library'`
- ✅  Unblocks [the TS migration of the rest of the `component-library` components](https://github.com/MetaMask/metamask-extension/issues/17725) as well as all extension TypeScript migration
- ✅  Improves the CSS specificity of the `Box` component to not override `ui/` and `app/` folder component styles like the current one does 

* Fixes #19239

### CSS Performance
This PR adds about `900kb` of extra CSS for the new `Box` component which I realize is A LOT. I've experimented using the same CSS for both components but that will either have some unexpected side effects if the old `Box` component uses the new CSS because of the updated specificity or we continue to have bad specificity for new `Box` component. I ran performance tests on css stylesheets and on average the increase in load time is around `79ms`. Performance tests should always be taken with a grain of salt because of background processes etc but overall I think it's not terrible and that we take this slight performance hit temporarily and actively start replacing the `Box` component through DS team and external contributor effort. Below is a table of 3 performance tests I ran on `develop` branch and this one.

| File             | Before (ms)              | Before Average (ms)     | After (ms)              | After Average (ms)     | Difference (ms)   |
|------------------|--------------------------|------------------------|-------------------------|------------------------|------------------:|
| index.css        | 214, 265, 301           | 260                    | 345, 301, 372          | 339.333                |       79.333      |
| index-rtl.css    | 267, 312, 581           | 386.667                | 306, 262, 329          | 299                    |      -87.667      |


![Screenshot 2023-06-07 at 11 34 24 AM](https://github.com/MetaMask/metamask-extension/assets/8112138/a299049f-1d6c-4922-b574-610f31883950)

![Screenshot 2023-06-07 at 11 39 46 AM](https://github.com/MetaMask/metamask-extension/assets/8112138/8f9c5fd4-8b36-4b22-a0bf-977fffe2d3ca)

## Next steps
Once this component has been merged we can deprecate the `Box` component in `ui/components/ui/box/box.js` and submit good first issues to the community to replace all instances of the old `Box` component inside the `component-library` folder and the rest of the code base. We can also update all the external contributor TS migration PRs

## Screenshots/Screencaps

### Before
Storybook of `Box` component in `ui/components/ui/box/box.js`

https://github.com/MetaMask/metamask-extension/assets/8112138/b71e98c8-a930-4ee5-8dab-82a15082f5de

Shows polymorphic types not working 

<img width="824" alt="Screenshot 2023-05-31 at 2 08 48 PM" src="https://github.com/MetaMask/metamask-extension/assets/8112138/d630d82d-833e-467d-8ebb-a2b0bb7fa43e">

### After
Storybook of `Box` component in `ui/components/component-library/box/box.tsx`

https://github.com/MetaMask/metamask-extension/assets/8112138/16d9028e-4a66-437d-82df-ebdc2284dc78

Shows polymorphic types working as expected and also erroring when not allowed

<img width="889" alt="Screenshot 2023-05-31 at 2 07 04 PM" src="https://github.com/MetaMask/metamask-extension/assets/8112138/985e9aae-e758-45fa-be0d-8b3328f69a79">

Testing Coverage

<img width="760" alt="Screenshot 2023-06-05 at 8 52 59 PM" src="https://github.com/MetaMask/metamask-extension/assets/8112138/9b342408-6c2a-4fea-8925-a43296402985">

## Manual Testing Steps
Storybook testing
- Go to the latest build of storybook in this PR
- Go to the component-library version of the `Box` component
- Check all stories

Polymorphic `as` prop testing
- Pull this branch 
- Navigate to the `ui/components/component-library/box/box.stories.tsx` story
- Play around with the `as` prop changing the root element and testing different html attributes associated with the element passed to the `as` prop to check type checking


## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
